File: wordcount_parte2/word_count_conc.py
import os
import sys
from threading import Thread, Semaphore

wc_count = 0
semaphore = Semaphore(1)

def wc(content):
    return len(content.split())

def wc_file(filename):
    try:
        with open(filename, 'r', encoding='latin-1') as f:
            file_content = f.read()
        return wc(file_content)
    except FileNotFoundError:
        return 0

# Função que cada Thread deve executar, e aqui ocorrerá região critica
def wc_subdir(dir_path):
    count = 0
    for filename in os.listdir(dir_path):
        filepath = os.path.join(dir_path, filename)
        count += wc_file(filepath)

    # Região crítica
    semaphore.acquire()
    global wc_count
    wc_count += count
    semaphore.release()

# Função que cria as Threads que executarão a sunção subdir
def wc_dir(dir_path):
    all_threads = []
    
    # Criando uma Thread para cada subdiretório
    for filename in os.listdir(dir_path):
        filepath = os.path.join(dir_path, filename)
        thread = Thread(target=wc_subdir, args=[filepath])
        all_threads.append(thread)
        thread.start()

    # Aguardando as threads terminarem
    for thread in all_threads:
        thread.join()

    # Imprimindo o resultado após todas as Threads terminarem
    global wc_count
    print(wc_count)

# Apenas inciando
def main():
    if len(sys.argv) != 2:
        print("Usage: python", sys.argv[0], "root_directory_path")
        return
    root_path = os.path.abspath(sys.argv[1])
    count = wc_dir(root_path)

if __name__ == "__main__":
    main()

File: wordcount_parte2/java/WordCountConc.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.concurrent.Semaphore;

public class WordCountConc {

    public static int WC = 0;

    public static Semaphore mutex= new Semaphore(1);

    public static void main(String[] args) throws InterruptedException {

        if (args.length != 1) {
            System.err.println("Usage: java WordCount <root_directory>");
            System.exit(1);
        }

        String rootPath = args[0];
        File rootDir = new File(rootPath);
        File[] subdirs = rootDir.listFiles();
        int numThreads = 0;

        for (File subdir : subdirs) {
            if (subdir.isDirectory()) {
                numThreads++;
            }
        }

        Thread[] threads = new Thread[numThreads];

        if (subdirs != null) {
            int pos = 0;
            for (File subdir : subdirs) {
                if (subdir.isDirectory()) {
                    String dirPath = rootPath + "/" + subdir.getName();
                    MyFunc run = new MyFunc(dirPath);
                    Thread t = new Thread(run);
                    threads[pos] = t;
                    t.start();
                }
                pos++;
            }

            for(Thread t : threads) {
                if(t != null){
                    t.join();
                }
            }
        }
        System.out.println(WC);
    }

    static class MyFunc implements Runnable {

        String dirpath;

        MyFunc(String dirpath) {this.dirpath = dirpath;}

        public void run() {
            try{
                wcDir(dirpath);
            }catch(Exception e){
                e.printStackTrace();
            }
        }

        private int wc(String fileContent) {
            String[] words = fileContent.split("\\s+");
            return words.length;
        }

        private int wcFile(String filePath) {
            try {
                BufferedReader reader = new BufferedReader(new FileReader(filePath));
                StringBuilder fileContent = new StringBuilder();
                String line;

                while ((line = reader.readLine()) != null) {
                    fileContent.append(line).append("\n");
                }

                reader.close();
                return wc(fileContent.toString());

            } catch (IOException e) {
                e.printStackTrace();
                return -1;
            }
        }

        private void wcDir(String dirPath) throws InterruptedException {
            File dir = new File(dirPath);
            File[] files = dir.listFiles();

            int count = 0;

            if (files != null) {
                for (File file : files) {
                    if (file.isFile()){
                        count += wcFile(file.getAbsolutePath());
                    }
                }
            }

            mutex.acquire();
            try{
                WC += count;
            } finally {
                mutex.release();
            }
        }

    }

}
